src/v/cloud_storage/remote_segment.cc

```diff
@@ -161,56 +168,101 @@ ss::future<> remote_segment::run_hydrate_bg() {
-                auto res = co_await _api.download_segment(
-                  _bucket, _path, _manifest, callback, local_rtc);
-                if (res != download_result::success) {
+            if (!_data_file) {
+                // We don't have a _data_file set so we have to check cache
+                // and retrieve the file out of it or hydrate.
+                // If _data_file is initialized we can use it safely since the
+                // cache can't delete it until we close it.
+                auto status = co_await _cache.is_cached(full_path);
+                switch (status) {
+                case cache_element_status::in_progress:
+                    vassert(
+                      false,
+                      "Hydration of segment {} is already in progress, {} "
+                      "waiters",
+                      full_path,
+                      _wait_list.size());
+                case cache_element_status::available:
                     vlog(
                       _ctxlog.debug,
-                      "Failed to hydrating a segment {}, {} waiter will be "
-                      "invoked",
+                      "Hydrated segment {} is already available, {} waiters "
+                      "will "
+                      "be invoked",
                       full_path,
                       _wait_list.size());
-                    err = std::make_exception_ptr(
-                      download_exception(res, full_path));
+                    break;
+                case cache_element_status::not_available: {
+                    vlog(_ctxlog.info, "Hydrating segment {}", full_path);
+                    auto callback =
+                      [this, full_path](
+                        uint64_t size_bytes,
+                        ss::input_stream<char> s) -> ss::future<uint64_t> {
+                        co_await _cache.put(full_path, s).finally([&s] {
+                            return s.close();
+                        });
+                        co_return size_bytes;
+                    };
+                    retry_chain_node local_rtc(
+                      cache_hydration_timeout, cache_hydration_backoff, &_rtc);
+                    auto res = co_await _api.download_segment(
+                      _bucket, _path, _manifest, callback, local_rtc);
+                    if (res != download_result::success) {
+                        vlog(
+                          _ctxlog.debug,
+                          "Failed to hydrating a segment {}, {} waiter will be "
+                          "invoked",
+                          full_path,
+                          _wait_list.size());
+                        err = std::make_exception_ptr(
+                          download_exception(res, full_path));
+                    }
+                } break;
+                }
+                if (!err) {
+                    auto maybe_file = co_await _cache.get(full_path);
+                    if (!maybe_file) {
```

Comment:
it is possible to do, but since it's not a hot path it's not necessary (it may also coerce the person who reads this code into thinking that it's called frequently)